apps/api/src/http/v3/rest/egvault/redeem.rs

```diff
@@ -0,0 +1,216 @@
+use shared::database::product::codes::{CodeEffect, RedeemCode};
+use shared::database::product::subscription::SubscriptionId;
+use shared::database::product::TimePeriod;
+use shared::database::queries::{filter, update};
+use shared::database::user::UserId;
+
+use super::{create_checkout_session_params, find_customer};
+use crate::global::Global;
+use crate::http::error::ApiError;
+use crate::http::middleware::auth::AuthSession;
+use crate::transactions::{with_transaction, TransactionError, TransactionResult, TransactionSession};
+
+pub async fn grant_entitlements(
+	tx: &mut TransactionSession<'_, ApiError>,
+	redeem_code: &RedeemCode,
+	user_id: UserId,
+) -> TransactionResult<(), ApiError> {
+	for (to, extends_subscription) in redeem_code.effects.iter().filter_map(|e| match e {
+		CodeEffect::Entitlement {
+			edge,
+			extends_subscription,
+		} => Some((edge, extends_subscription)),
+		_ => None,
+	}) {
+		let from = match extends_subscription {
+			Some(extends_subscription) => EntitlementEdgeKind::Subscription {
+				subscription_id: SubscriptionId {
+					user_id: user_id,
+					product_id: *extends_subscription,
+				},
+			},
+			None => EntitlementEdgeKind::User { user_id },
+		};
+
+		tx.insert_one(
+			EntitlementEdge {
+				id: EntitlementEdgeId {
+					from,
+					to: to.clone(),
+					managed_by: Some(EntitlementEdgeManagedBy::RedeemCode {
+						redeem_code_id: redeem_code.id,
+					}),
+				},
+			},
+			None,
+		)
+		.await?;
+	}
+
+	Ok(())
+}
+
+#[derive(Debug, serde::Deserialize)]
+pub struct RedeemRequest {
+	code: String,
+}
+
+#[derive(Debug, serde::Serialize)]
+pub struct RedeemResponse {
+	/// Url that the website will open
+	authorize_url: Option<String>,
+	/// list of ids of cosmetics that the user received
+	items: Vec<String>,
+}
+
+pub async fn redeem(
+	State(global): State<Arc<Global>>,
+	auth_session: Option<Extension<AuthSession>>,
+	Json(body): Json<RedeemRequest>,
+) -> Result<Json<RedeemResponse>, ApiError> {
+	let auth_session = auth_session.ok_or(ApiError::UNAUTHORIZED)?;
+
+	let res = with_transaction(&global, |mut tx| {
+		let global = Arc::clone(&global);
+
+		async move {
+			let code = tx
+				.find_one_and_update(
+					filter::filter! {
+						RedeemCode {
+							code: body.code,
+							#[query(selector = "gt")]
+							remaining_uses: 0,
+							#[query(flatten)]
+							active_period: TimePeriod {
+								#[query(selector = "lt")]
+								start: chrono::Utc::now(),
+								#[query(selector = "gt")]
+								end: chrono::Utc::now(),
+							},
+						}
+					},
+					update::update! {
+						#[query(inc)]
+						RedeemCode {
+							remaining_uses: -1,
+						},
+					},
+					None,
+				)
+				.await?
+				.ok_or(TransactionError::custom(ApiError::NOT_FOUND))?;
+
+			// if the user is already subscribed
+			let is_subscribed = auth_session
+				.user(&global)
+				.await
+				.map_err(TransactionError::custom)?
+				.computed
+				.is_subscribed;
```

Comment:
I think what we do here is write a query to check if any period is active in which case they are subbed.